src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -42,19 +42,24 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             var indexedDirectDependenciesKeyNames = new Lazy<HashSet<string>>(
                 () =>
                 {
-                    var result = new HashSet<string>();
+                    var result = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                     result.AddRange(rootNode.InnerNodes.Select(n => n.Key.Name));
                     return result;
                 });
 
-            while (transitiveCentralPackageVersions.TryTake(out LibraryDependency centralPackageversionDependecy))
+            var transitiveCentralPackageVersionNodes = new List<GraphNode<RemoteResolveResult>>();
+            while (transitiveCentralPackageVersions.TryTake(out LibraryDependency centralPackageVersionDependecy))
             {
                 // do not add a transitive dependency node if it is direct already
-                if (!indexedDirectDependenciesKeyNames.Value.Contains(centralPackageversionDependecy.Name))
+                if (!indexedDirectDependenciesKeyNames.Value.Contains(centralPackageVersionDependecy.Name))
                 {
-                    await AddTransitiveCentralPackageVersionNodesAsync(rootNode, centralPackageversionDependecy, framework, runtimeIdentifier, runtimeGraph, transitiveCentralPackageVersions);
+                    // as the nodes are created more parents can be added for a single central transitive node
+                    // keep the list of the nodes created and add the parents's references at the end
+                    // the parent references are needed to keep track of possible rejected parents
+                    transitiveCentralPackageVersionNodes.Add(await AddTransitiveCentralPackageVersionNodesAsync(rootNode, centralPackageVersionDependecy, framework, runtimeIdentifier, runtimeGraph, transitiveCentralPackageVersions));
```

Comment:
Wouldn't it be more efficient to add all the discovered conflicts as top level and then call CreateGraphNode again? 

Specifically say you have the following: 

Packages Available on the feed: 
```
A 1.0.0 -> B 1.0.0
A 1.0.0 -> C 1.0.0
B 1.0.0
B 2.0.0 -> C 1.0.0
C 1.0.0
C 2.0.0 -> D 1.0.0 
D 1.0.0
```

The project has the following reference:
```
A
```

The central file has:

```
A 1.0.0
B 2.0.0
C 2.0.0
```

Now you run phase 1, you detect that B & C need bumping. You currently add them 1 by 1. 

Test example: 
```cs
/// <summary>
        /// Start -> A 1.0.0.
        ///
        /// Central file has:
        /// A 1.0.0
        /// B 2.0.0
        /// C 2.0.0
        /// D 2.0.0
        ///
        /// The package dependencies are:
        /// A 1.0.0 -> B 1.0.0
        /// B 2.0.0 -> C 1.0.0
        /// C 2.0.0 -> D 1.0.0
        ///
        /// All mentioned versions are available on the feeds
        /// </summary>
        [Fact]
        public async Task WalAsync_WithMultiLevelCentralTransitiveDependencies_ResolvesToCentrallySpecifiedVersion()
        {
            var framework = NuGetFramework.Parse("net45");

            var context = new TestRemoteWalkContext();
            var provider = new DependencyProvider();
            var version100 = "1.0.0";
            var version200 = "2.0.0";

            // A 1.0.0 -> B 1.0.0
            // A 1.0.0 -> C 1.0.0
            provider.Package("A", version100)
                    .DependsOn("B", version100)
                    .DependsOn("C", version100);

            // B 1.0.0
            provider.Package("B", version100);

            // B 2.0.0 -> C 1.0.0
            provider.Package("B", version200)
                   .DependsOn("C", version100);

            // C 1.0.0
            provider.Package("C", version100);

            // C 2.0.0 -> D 1.0.0
            provider.Package("C", version200)
                   .DependsOn("D", version100);

            provider.Package("C", version200);

            provider.Package("Start", version100)
                     .DependsOn("A", version100, LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.Direct)
                     .DependsOn("B", version200, LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.None)
                     .DependsOn("C", version200, LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.None)
                     .DependsOn("D", version200, LibraryDependencyTarget.Package, versionCentrallyManaged: true, libraryDependencyReferenceType: LibraryDependencyReferenceType.None);

            context.LocalLibraryProviders.Add(provider);
            var walker = new RemoteDependencyWalker(context);

            // Act
            var rootNode = await DoWalkAsync(walker, "Start", framework);

            // Assert
            Assert.Equal(4, rootNode.InnerNodes.Count);

            AnalyzeResult<RemoteResolveResult> result = rootNode.Analyze();

            rootNode.InnerNodes.Where(n => n.Item.IsCentralTransitive).Should().HaveCount(3);
        }
```

I think this is the perf scenario that we discussed with @davidfowl.